File: frontend/src/lib/roles.ts
import { useAuthStore } from '../store/auth'
import type { UserRole } from '../types/api'

/** Numeric rank: higher = more privileged. */
export const roleRank: Record<UserRole, number> = {
  viewer: 0,
  operator: 1,
  admin: 2,
}

/**
 * Returns true when `userRole` meets or exceeds `min`.
 * Unknown roles (empty string / undefined) rank as viewer (0).
 */
export function hasRole(userRole: string | undefined | null, min: UserRole): boolean {
  const rank = roleRank[(userRole as UserRole) ?? 'viewer'] ?? 0
  return rank >= roleRank[min]
}

export interface CanResult {
  /** True for operator-or-higher (start/stop/restart containers, bulk start/stop/restart). */
  isOperator: boolean
  /** True for admin only (secrets, scripts, webhooks, user management, destructive ops). */
  isAdmin: boolean
}

/** Hook returning role-derived capability flags for the currently signed-in user. */
export function useCan(): CanResult {
  const user = useAuthStore((s) => s.user)
  const role = user?.role
  return {
    isOperator: hasRole(role, 'operator'),
    isAdmin: hasRole(role, 'admin'),
  }
}

File: backend/auth/roles.go
package auth

// Roles, lowest to highest privilege (Feature 30). The hierarchy is:
//
//	viewer    — read-only access to inventory, files, logs, metrics.
//	operator  — viewer + non-destructive container lifecycle (start/stop/restart).
//	admin      — full access, including destructive ops, secrets, and user mgmt.
//
// The Auth middleware reloads the user from the DB on every request, so a role
// change takes effect on the user's next call — no token revocation is needed
// to enforce a downgrade.
const (
	RoleViewer   = "viewer"
	RoleOperator = "operator"
	RoleAdmin    = "admin"
)

var roleRank = map[string]int{
	RoleViewer:   1,
	RoleOperator: 2,
	RoleAdmin:    3,
}

// RoleRank returns the privilege rank of a role, or 0 for an unknown role.
func RoleRank(role string) int { return roleRank[role] }

// RoleValid reports whether role is one of the known roles.
func RoleValid(role string) bool {
	_, ok := roleRank[role]
	return ok
}

// AtLeast reports whether role meets or exceeds min. An unknown role (rank 0)
// never qualifies, so a malformed/empty role is fail-closed.
func AtLeast(role, min string) bool {
	r, m := roleRank[role], roleRank[min]
	return r != 0 && m != 0 && r >= m
}
